### PR TITLE
Allow custom base class as node customization point

### PR DESCRIPTION
**Pull request by barcode**
This PR adds an additional template parameter which allows to set a custom base class for `nlohmann::json`. This class serves as an extension point and allows to add functionality to json node. Examples for such functionality might be metadata or additional member functions (e.g. visitors) or other application specific code.
By default the parameter is set to void and an empty base class is used. In this case the library behaves as it already did.

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src). (should be. will update state aver coverall is done)
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

### CODE REVIEW

**Comment by barcode** [comment]
Just saw i used the wrong base branch and fixed it.

**Comment by nlohmann** [comment]
Not sure about the cppcheck error, but there are also issues with MSVC.

**Comment by nlohmann** [comment]
Great! Windows seem to work now. The amalgamation check failed though. Please try again after calling `make amalgamate`.

**Comment by barcode** [comment]
For some strange reason `make amalgamate` did not do anything (`make: Nothing to be done for 'amalgamate'.`). After deleting the single include and running `make amalgamate` again, it also did some formatting. Not sure if that was supposed to happen.

The [cppcheck test](https://github.com/nlohmann/json/runs/4056540972?check_suite_focus=true) was skipped this time, so it did not fail. I guess it will fail again. 
Not sure why it is complaining about, but it had the same issue on my old PR

    Check code with Cppcheck
    Checking /__w/json/json/single_include/nlohmann/json.hpp ...
    /__w/json/json/single_include/nlohmann/json.hpp:3448:54: error: syntax error [syntaxError]
             class BinaryType = std::vector<std::uint8_t>,
                                                         ^

The ci uses cppcheck 2.7. This seems to be some version of the current main branch. If i build cppcheck main and use it, i get the same error. It looks like a false positive. I could suppress it using `// cppcheck-suppress syntaxError`.

**Comment by nlohmann** [comment]
> For some strange reason `make amalgamate` did not do anything (`make: Nothing to be done for 'amalgamate'.`). After deleting the single include and running `make amalgamate` again, it also did some formatting. Not sure if that was supposed to happen.

Unfortunately, this can happen. You fixed it the right way. Sorry for the inconvenience.

> The ci uses cppcheck 2.7. This seems to be some version of the current main branch. If i build cppcheck main and use it, i get the same error. It looks like a false positive. I could suppress it using `// cppcheck-suppress syntaxError`.

Yes, the CI uses the bleeding-edge version of cppcheck. If you can suppress the issue, then I guess this is the way.

**Comment by barcode** [comment]
Finally all checks went green. All but the first commit were fixes for the CI and do not carry important information. I think merge+squash would be a good way to keep the commit history  in a readable form.

**Comment by nlohmann** [comment]
`std::make_unique` is not available in C++11 - I guess that's why the CI fails.

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp:12]: Please add a document why you need `json_base_class`.

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp:3]:
```suggestion
#include <type_traits> // conditional, is_same
```

File: test/src/unit-custom-base-class.cpp
Comment (nlohmann) [inline test/src/unit-custom-base-class.cpp:204]: Please document the scenario you want to test here.

File: test/src/unit-custom-base-class.cpp
Comment (nlohmann) [inline test/src/unit-custom-base-class.cpp:38]: Please document the scenario you want to test here.

**Review by nlohmann — CHANGES_REQUESTED** [review]

File: test/src/unit-custom-base-class.cpp
Comment (barcode) [reply test/src/unit-custom-base-class.cpp:40]: see 596e278

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (barcode) [reply include/nlohmann/detail/json_custom_base_class.hpp:29]: see 596e278

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp:20]: This paragraph seems to be a duplicate of the previous one.

File: test/src/unit-custom-base-class.cpp
Comment (nlohmann) [inline test/src/unit-custom-base-class.cpp:39]:
```suggestion
// Add some metadata to each node and test the behaviour of copy / move
```

**Review by nlohmann — CHANGES_REQUESTED** [review]

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (barcode) [reply include/nlohmann/detail/json_custom_base_class.hpp]: Ups. Fixed in 3311d11

**Comment by gregmarr** [comment]
@barcode @nlohmann I think this just needs some documentation.

File: docs/examples/json_base_class_t.cpp
Comment (falbrechtskirchinger) [inline docs/examples/json_base_class_t.cpp:68]: I'd put `case value_t::discarded` and `default` at the end of the switch.

```cpp
case value_t::discarded:
default:
    break;
```

File: include/nlohmann/json.hpp
Comment (falbrechtskirchinger) [inline include/nlohmann/json.hpp:1241]: Please keep `std::move` and use `NOLINT` instead.
```cpp
: json_base_class_t(std::move(other)), // NOLINT(misc-use-after-move)
```

File: include/nlohmann/json_fwd.hpp
Comment (falbrechtskirchinger) [inline include/nlohmann/json_fwd.hpp:39]: Can you please check if this is still needed for cppcheck 2.8? I believe CI is using this version.

File: tests/src/unit-custom-base-class.cpp
Comment (falbrechtskirchinger) [inline tests/src/unit-custom-base-class.cpp:71]: Yikes. Should we consider adding a template alias to simplify the default use cases?

Something like this:
```cpp
template<typename CustomBaseClass, ...>
using json_with_base = basic_json<..., CustomBaseClass>;
```

(Maybe in a separate PR.)

File: tests/src/unit-custom-base-class.cpp
Comment (falbrechtskirchinger) [inline tests/src/unit-custom-base-class.cpp:178]: Please add the rule in parentheses.

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (falbrechtskirchinger) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md:24]:
```suggestion
`basic_json` only supports  copy/move construction/assignement if `CustomBaseClass` does so as well.
```

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (falbrechtskirchinger) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md:21]:
```suggestion
#### Limitations
```
Subtle difference.

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (falbrechtskirchinger) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md:8]:
```suggestion
Examples of such functionality might be metadata, additional member functions (e.g. visitors), or other application-specific code.
```

**Review by falbrechtskirchinger — CHANGES_REQUESTED** [review]

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (gregmarr) [reply docs/mkdocs/docs/api/basic_json/json_base_class_t.md:24]: Typo: assignement should be assignment.

File: tests/src/unit-custom-base-class.cpp
Comment (barcode) [reply tests/src/unit-custom-base-class.cpp:71]: I agree, the current syntax is ugly.
If we want to add an alias, it would always have to be consistent with the template for `basic_json`

e.g.:

    template<class CustomBaseClass,
            template<typename U, typename V, typename... Args> class ObjectType =
            std::map,
            template<typename U, typename... Args> class ArrayType = std::vector,
            class StringType = std::string, class BooleanType = bool,
            class NumberIntegerType = std::int64_t,
            class NumberUnsignedType = std::uint64_t,
            class NumberFloatType = double,
            template<typename U> class AllocatorType = std::allocator,
            template<typename T, typename SFINAE = void> class JSONSerializer =
            adl_serializer,
            class BinaryType = std::vector<std::uint8_t>>
    using json_with_base = basic_json<ObjectType, ArrayType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass>;

So we always would have to duplicate and update everything from the base template.

There is a way to circumvent this. We could add `basic_json::with_base_t<T>`. This alias would use the parameters from the current type. In this case the namespace level alias could be:

    template<class CustomBaseClass, class...Ts>
    using json_with_base = basic_json<Ts...>::with_base_t<CustomBaseClass>;

Now we would only have to update `basic_json::with_base_t` and not the defaults.

Since this change is not trivial and might need some discussion, i would leave it to a different PR.

File: include/nlohmann/json.hpp
Comment (barcode) [reply include/nlohmann/json.hpp]: Done

File: include/nlohmann/json_fwd.hpp
Comment (barcode) [reply include/nlohmann/json_fwd.hpp]: It is not required anymore. I removed it.

File: tests/src/unit-custom-base-class.cpp
Comment (barcode) [reply tests/src/unit-custom-base-class.cpp]: Added

**Comment by barcode** [comment]
Strange... not sure why the previous run of cppcheck accepted the code without the suppress comment. I guess it is a sporadic error. Hence i added the comment again.

**Review by shaischaudhry — APPROVED** [review]

**Comment by gregmarr** [comment]
@barcode This needs to be updated for the new type safety namespace introduced in 3.11.0 through 3.11.2.

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (falbrechtskirchinger) [inline include/nlohmann/detail/json_custom_base_class.hpp]: Please use `NLOHMANN_JSON_NAMESPACE_BEGIN` to open the `nlohmann` namespace here and `NLOHMANN_JSON_NAMESPACE_END` at the end of the file to close it.

```suggestion
NLOHMANN_JSON_NAMESPACE_BEGIN
```

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (falbrechtskirchinger) [inline include/nlohmann/detail/json_custom_base_class.hpp]:
```suggestion
NLOHMANN_JSON_NAMESPACE_END
```

File: single_include/nlohmann/json.hpp
Comment (falbrechtskirchinger) [inline single_include/nlohmann/json.hpp]:
```suggestion
NLOHMANN_JSON_NAMESPACE_BEGIN
```

File: single_include/nlohmann/json.hpp
Comment (falbrechtskirchinger) [inline single_include/nlohmann/json.hpp]:
```suggestion
NLOHMANN_JSON_NAMESPACE_END
```

**Review by falbrechtskirchinger — CHANGES_REQUESTED** [review]
If you apply all four suggestions through GitHub you shouldn't have to re-amalgamate the headers and the failing checks should pass.

**Comment by barcode** [comment]
So what is the Plan on this PR? Is there something still to be done by me?
If not, what is the plan for merging this PR?
i just want to know, since i will have to rebase #3165 at some point and #3165 requires this PR to work.

File: tests/src/unit-custom-base-class.cpp
Comment (falbrechtskirchinger) [reply tests/src/unit-custom-base-class.cpp:71]: @barcode You can mark this as resolved.

**Comment by falbrechtskirchinger** [comment]
Looks like it just needs another review (which I can do either a little later today or over the weekend) and some attention from @nlohmann.

**Comment by nlohmann** [comment]
I'll check during the weekend.

**Review by falbrechtskirchinger — APPROVED** [review]
Looks good to me.

**Comment by falbrechtskirchinger** [comment]
Some notes for a follow-up PR:

* Make specifying a base class easier/less verbose.
* Warn users about potential conflicts with newer versions, i.e., base members should avoid generic names to avoid clashes with future member additions to `basic_json`.
* Doxygen comments should conform to the "new" convention.
* (Run reuse.)

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (nlohmann) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md]:
```suggestion
Examples of such functionality might be metadata, additional member functions (e.g., visitors), or other application-specific code.
```

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (nlohmann) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md]:
```suggestion
The type `CustomBaseClass` has to be a default-constructible class.
```

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (nlohmann) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md]:
```suggestion
`basic_json` only supports copy/move construction/assignment if `CustomBaseClass` does so as well.
```

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (nlohmann) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md]:
```suggestion
- Added in version 3.12.0.
```

File: docs/mkdocs/docs/api/basic_json/json_base_class_t.md
Comment (nlohmann) [inline docs/mkdocs/docs/api/basic_json/json_base_class_t.md]:
```suggestion
The default value for `CustomBaseClass` is `void`. In this case an [empty base class](https://en.cppreference.com/w/cpp/language/ebo) is used and no additional functionality is injected.
```

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp]:
```suggestion
of @ref basic_json do not require complex case distinctions
```

**Review by nlohmann — CHANGES_REQUESTED** [review]
Some minor comments, but otherwise looks good to me.

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp]:
```suggestion
So that the correct implementations of the copy / move ctors / assign operators
```

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Will merge once the CI is green.

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (nlohmann) [inline include/nlohmann/detail/json_custom_base_class.hpp:7]: The CI complains that this symbol cannot be found.

Please include `<nlohmann/detail/abi_macros.hpp>`.

**Review by nlohmann — CHANGES_REQUESTED** [review]

File: include/nlohmann/detail/json_custom_base_class.hpp
Comment (barcode) [reply include/nlohmann/detail/json_custom_base_class.hpp:7]: Added

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Thanks a lot!!!

**Comment by barcode** [comment]
Thanks to you guys, too. I really like this library and know it is a lot of work to maintain libraries.

File: include/nlohmann/json.hpp
Comment (nlohmann) [inline include/nlohmann/json.hpp:1207]: @barcode Sorry for being so late, but now I got a warning from Clang-Tidy  about accessing `other.m_value` after moving `other` in the line before. Any idea how to fix this?

**Pull request merged**
